extensions/barrage/src/main/java/io/deephaven/extensions/barrage/util/BarrageStreamReader.java

```diff
@@ -151,8 +151,34 @@ public BarrageMessage safelyParseFrom(final StreamReaderOptions options,
                     throw new IllegalStateException("Only know how to decode Schema/BarrageRecordBatch messages");
                 }
 
+                // snapshots may not provide metadata, generate it now
                 if (msg == null) {
-                    throw new IllegalStateException("Could not detect barrage metadata for update");
+                    msg = new BarrageMessage();
+
+                    // generate a default set of column selectors
+                    msg.snapshotColumns = new BitSet(columnTypes.length);
+
+                    // create and fill the add column metadata from the schema
+                    msg.addColumnData = new BarrageMessage.AddColumnData[columnTypes.length];
+                    for (int ci = 0; ci < msg.addColumnData.length; ++ci) {
+                        msg.addColumnData[ci] = new BarrageMessage.AddColumnData();
+                        msg.addColumnData[ci].type = columnTypes[ci];
+                        msg.addColumnData[ci].componentType = componentTypes[ci];
+
+                        // set the column selector active by default
+                        msg.snapshotColumns.set(ci);
+                    }
+
+                    // no mod column data
+                    msg.modColumnData = new BarrageMessage.ModColumnData[0];
+
+                    // generate empty row sets
+                    msg.rowsRemoved = RowSetFactory.empty();
+                    msg.shifted = new RowSetShiftData.Builder().build();
```

Comment:
There's an EMPTY singleton for this. (`io.deephaven.engine.rowset.RowSetShiftData#EMPTY`)